src/bpfilter/xlate/cli.c

```diff
@@ -56,6 +66,203 @@ int _bf_cli_ruleset_flush(const struct bf_request *request,
+                if (r < 0)
+                    return bf_err_r(r, "failed to get rule counter\n");
+            }
+
+            /* Note: The map has entries for every rule regardless of whether the
+             * input rule specified "COUNTERS" or not. Thus, advancing map_index
+             * is unconditional. When rule->counters == false, the corresponding
+             * counter value will remain {0,0}. */
+            counter_index++;
+            map_idx++;
+        }
+
+        // Chain's policy counter
+        r = bf_cgen_get_counter(cgen, BF_COUNTER_POLICY,
+                                &counters[counter_index]);
+        if (r < 0)
+            return bf_err_r(r, "failed to get policy counter\n");
+        counter_index++;
+
+        // Chain's error counter
+        r = bf_cgen_get_counter(cgen, BF_COUNTER_ERRORS,
+                                &counters[counter_index]);
+        if (r < 0)
+            return bf_err_r(r, "failed to get error counter\n");
+    }
+
+    return 0;
+}
+
+/**
+ * Populate a list of chains from list of code generators.
+ *
+ * @param cgens Pointer to list of generators. Must be non-NULL.
+ * @param chains Pointer to list for storing the chains. Must be non-NULL.
+ * @return 0 on success or negative error code on failure.
+ */
+static int _bf_cli_get_chain_list(bf_list *cgens, bf_list *chains)
+{
+    _clean_bf_list_ bf_list _chains = bf_list_default(NULL, bf_chain_marsh);
+    int r;
+
+    bf_assert(cgens && chains);
+
+    // iterate over the codegens to get the chains
+    bf_list_foreach (cgens, cgen_node) {
+        struct bf_cgen *cgen = bf_list_node_get_data(cgen_node);
+        r = bf_list_add_tail(&_chains, cgen->chain);
+        if (r)
+            return bf_err_r(r, "failed to add chain to list");
+    }
+
+    *chains = bf_list_move(_chains);
+
+    return 0;
+}
+
+/**
+ * Create a marshalled structure of counters.
+ *
+ * @param cgens A list of code generators. Must be non-NULL.
+ * @param counter_marsh A pointer to a bf_marsh pointer where the counters will be marshalled. Must be non-NULL.
+ * @return 0 on success or negative error code on failure.
+ */
+static int _bf_cli_get_counters_marsh(bf_list *cgens,
+                                      struct bf_marsh **counter_marsh)
+{
+    int r;
+    size_t num_counters = 0;
+    _cleanup_free_ struct bf_counter *counters = NULL;
+
+    bf_assert(counter_marsh && cgens);
+
+    /* Each chain has a policy counter and an error counter.
+     * Each rule has a counter, though it may be unused. */
+    num_counters = (2 * bf_list_size(cgens)) + _bf_cli_num_rules(cgens);
+
+    counters = calloc(num_counters, sizeof(struct bf_counter));
+    if (!counters)
+        return bf_err_r(-ENOMEM, "failed to allocate memory for counters\n");
+
+    r = _bf_cli_get_ctr_vals(cgens, counters);
+    if (r < 0)
+        return bf_err_r(r, "could not get ctr vals\n");
+
+    size_t marsh_size = num_counters * sizeof(struct bf_counter);
```

Comment:
Don't overthink it:
```
_clean_bf_list_ bf_list counters = bf_list_default(bf_counter_free, bf_counter_marsh);

foreach (chain) {
    foreach (rule) {
        _cleanup_free_ struct bf_counter *counter = NULL;
        counter = malloc(sizeof(*counter));
        // check for NULL

       bf_cgen_get_counter(..., counter);
       bf_list_add(&counters, counter);
       TAKE_PTR(counter);
    }
}

bf_list_marsh(counters);
```

That way you don't worry about the size or the serialization of the data.